scalyr_agent/builtin_monitors/kubernetes_monitor.py

```diff
@@ -2281,14 +2281,17 @@ def _get_containers(
                         )
                     except k8s_utils.K8sApiException as e:
                         # If the pod details cannot be retrieved from the K8s API:
-                        # 404 - log warning
+                        # 404 - log warning, include based on k8s_include_by_default
                         # Otherwise (401 as per K8s Api documentation) - log error
                         # Exclude the pod
                         if e.status_code == 404:
-                            global_log.warning(
-                                "Pod %s/%s not found in K8s API. Excluding pod from collection."
-                                % (pod_namespace, pod_name),
-                                exc_info=e,
+                            if k8s_include_by_default:
+                                include_message = "Including pod based on SCALYR_K8S_INCLUDE_ALL_CONTAINERS=true."
+                            else:
+                                include_message = "Excluding pod based on SCALYR_K8S_INCLUDE_ALL_CONTAINERS=false."
```

Comment:
Just to make sure I understand it correctly, specific (short lived) container will be included if it was already in cache, even if we had `exclude: true` annotation on the pod / container, right (since it may already disappear from the API)?

If that is indeed the case, it may be a good idea to document this somewhere.

And if that is not the case, we should have an end to end test case which exercises that (pod / container which has `exclude: true` annotation and verifying that those logs are not included).

I also wonder if there could be some yet undetermined negative consequences of this approach - e.g. it may result in more pod / container logs to be included under some edge cases and not just short lived ones.

Perhaps it would be a good idea to deploy it to some larger testing environment with `SCALYR_K8S_INCLUDE_ALL_CONTAINERS=true` and some pods having set `exlcude: true` and observe how it behaves.

Reply:
Yes, the requirement was to include it even though it might have been annotated by "exclude: true".
The idea behind it is to be on the safe side and rather log something unnecessarily than lose something you need.
Long term testing it is a good idea. In this case I am reverting to a previous state of things and need to release it soon, but my plan is to deploy and test any change for at least a week to a testing or staging env. I will include some of the already release, including this one.
And I will write the doc.